addons/pkg/util/cluster_util.go

```diff
@@ -163,3 +165,42 @@ func GetClusterLabel(clusterLabels map[string]string, labelKey string) string {
 
 	return labelValue
 }
+
+// GetInfraProvider get infrastructure kind from cluster spec
+func GetInfraProvider(cluster *clusterv1beta1.Cluster) (string, error) {
+	var infraProvider string
+
+	if cluster.Spec.InfrastructureRef != nil {
+		infraProvider = cluster.Spec.InfrastructureRef.Kind
+		switch infraProvider {
+		case tkgconstants.InfrastructureRefVSphere:
+			return tkgconstants.InfrastructureProviderVSphere, nil
+		case tkgconstants.InfrastructureRefAWS:
+			return tkgconstants.InfrastructureProviderAWS, nil
+		case tkgconstants.InfrastructureRefAzure:
+			return tkgconstants.InfrastructureProviderAzure, nil
+		case tkgconstants.InfrastructureRefDocker:
+			return tkgconstants.InfrastructureProviderDocker, nil
+		}
+	}
+
+	return "", errors.New("unknown error in getting infraProvider")
+}
+
+// IsTKGSCluster checks if the cluster is a TKGS cluster
+func IsTKGSCluster(ctx context.Context, c client.Client, cluster *clusterv1beta1.Cluster) (bool, error) {
+	// Verify if operating on a TKGS cluster by checking if virtualmachine objects exist for the cluster label
+	virtualMachineList := &vmoperatorv1alpha1.VirtualMachineList{}
+	listOptions := client.MatchingLabels{
+		tkgconstants.CAPVClusterSelectorKey: cluster.Name,
+	}
+
+	if err := c.List(ctx, virtualMachineList, listOptions); err != nil {
+		// If CRD resource doesn't exist on cluster, it will throw an unKnownError with the error message that contains `no matches for kind "VirtualMachine"`
```

Comment:
I probably didn't catch this earlier but this is brittle. I would use the capability client to check if resource is available, see https://github.com/vmware-tanzu/tanzu-framework/blob/main/cmd/cli/plugin/package/main.go#L72 for usage in package client.

Reply:
I added some custom logic using discovery client. 

The reason why I didn't use capability library is because we are using `memCache` discovery client. It will simply return a plain text error `not found`, if Group doesn't exist in the cluster. https://github.com/kubernetes/client-go/blob/master/discovery/cached/memory/memcache.go#L98

While capability library library is trying to catch `apierrors.IsNotFound` https://github.com/vmware-tanzu/tanzu-framework/blob/main/pkg/v1/sdk/capabilities/discovery/cluster_gvr.go#L221-L223.  Since the error we got is not an IsNotFound api error, it will be returned as an error. However, we want the error to be nil if the resource is not found.

Reply:
A possible workaround to use capability library is by adding `strings.Contains(err.Error(), "not found")` into https://github.com/vmware-tanzu/tanzu-framework/blob/main/pkg/v1/sdk/capabilities/discovery/cluster_gvr.go#L221-L223. But I think we should be fine with adding some custom logic instead of using capability library.